go/test/endtoend/vreplication/materialize_test.go

```diff
@@ -96,3 +96,144 @@ func TestShardedMaterialize(t *testing.T) {
+`
+
+const smMaterializeVSchemaSource = `
+{
+  "sharded": true,
+  "tables": {
+      "mat": {
+          "column_vindexes": [
+              {
+                  "column": "id",
+                  "name": "hash"
+              }
+          ]
+      }
+  },
+  "vindexes": {
+      "hash": {
+          "type": "hash"
+      }
+  }
+}
+`
+const smMaterializeSchemaTarget = `
+	CREATE TABLE mat2 (
+	id bigint NOT NULL,
+	val varbinary(10) NOT NULL,
+	ts timestamp NOT NULL,
+	day int NOT NULL,
+	month int NOT NULL,
+    x int not null,
+	PRIMARY KEY (id)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
+`
+
+const smMaterializeVSchemaTarget = `
+{
+  "sharded": true,
+  "tables": {
+      "mat2": {
+          "column_vindexes": [
+              {
+                  "column": "id",
+                  "name": "hash"
+              }
+          ]
+      }
+  },
+  "vindexes": {
+      "hash": {
+          "type": "hash"
+      }
+  }
+}
+`
+const smMaterializeSpec2 = `{"workflow": "wf1", "source_keyspace": "source", "target_keyspace": "target", "table_settings": [ {"target_table": "mat2", "source_expression": "select id, val, ts, dayofmonth(ts) as day, month(ts) as month, custom1(id, val) as x from mat"  }] }`
+
+const materializeInitDataQuery = `insert into mat(id, val, ts) values (1, 'abc', '2021-10-9 16:17:36'), (2, 'def', '2021-11-10 16:17:36')`
+
+const customFunc = `
+CREATE FUNCTION custom1(id int, val varbinary(10))
+RETURNS int
+DETERMINISTIC
+RETURN id * length(val);
+`
+
+// TestMaterialize: details mentioned above
+func testMaterialize(t *testing.T) {
+	defaultCellName := "zone1"
+	allCells := []string{"zone1"}
+	allCellNames = "zone1"
+	vc = NewVitessCluster(t, "TestMaterialize", allCells, mainClusterConfig)
+	sourceKs := "source"
+	targetKs := "target"
+	require.NotNil(t, vc)
+	defaultReplicas = 0 // because of CI resource constraints we can only run this test with primary tablets
+	defer func() { defaultReplicas = 1 }()
+
+	defer vc.TearDown(t)
+
+	defaultCell = vc.Cells[defaultCellName]
+	vc.AddKeyspace(t, []*Cell{defaultCell}, sourceKs, "-", smMaterializeVSchemaSource, smMaterializeSchemaSource, defaultReplicas, defaultRdonly, 300)
+	vtgate = defaultCell.Vtgates[0]
+	require.NotNil(t, vtgate)
+	vtgate.WaitForStatusOfTabletInShard(fmt.Sprintf("%s.%s.primary", sourceKs, "0"), 1)
+
+	vc.AddKeyspace(t, []*Cell{defaultCell}, targetKs, "-", smMaterializeVSchemaTarget, smMaterializeSchemaTarget, defaultReplicas, defaultRdonly, 400)
+	vtgate.WaitForStatusOfTabletInShard(fmt.Sprintf("%s.%s.primary", targetKs, "0"), 1)
+
+	vtgateConn = getConnection(t, vc.ClusterConfig.hostname, vc.ClusterConfig.vtgateMySQLPort)
+	defer vtgateConn.Close()
+	verifyClusterHealth(t, vc)
+
+	_, err := vtgateConn.ExecuteFetch(materializeInitDataQuery, 0, false)
+	require.NoError(t, err)
+
+	ks2Keyspace := vc.Cells[defaultCell.Name].Keyspaces[targetKs]
+	ks2Primary := ks2Keyspace.Shards["-"].Tablets["zone1-400"].Vttablet
```

Comment:
I think we can do this instead of relying on a hardcoded tablet UID: 
```
ks2Primary := vc.getVttabletsInKeyspace(t, defaultCellName, targetKs, "primary")
```

Reply:
`getVttabletsInKeyspace` returns a map with tablet name as the key. In this case we get a map with a single entry since we only have one shard and hence only one primary, but in general we could have multiple entries for each shard. I added another helper `getPrimaryTablet(keyspace, shard)` for this purpose.